### PR TITLE
Refract member element - reimplementation in way "key"/"value"

### PR DESCRIPTION
**Pull request by klokane**
Ready for review, please check especially result of parsing for fixtures:
`test/fixtures/mson-*`

### CODE REVIEW

File: src/RefractAST.cc [hunk 1]
```diff
@@ -147,8 +147,9 @@ namespace drafter
     };
 
 
-    std::vector<refract::IElement*> MsonElementsToRefract(const mson::Elements& elements) {
-        std::vector<refract::IElement*> result;
+    RefractElements MsonElementsToRefract(const mson::Elements& elements) {
+        //std::vector<refract::IElement*> result;
```
Comment (zdne) [inline src/RefractAST.cc]: Why the commented out code?

File: src/RefractAST.cc [hunk 2]
```diff
@@ -315,36 +327,82 @@ namespace drafter
...
-            if (IElement* e = SimplifyRefractContainer(samples)) {
-                element->attributes["sample"] = e;
-            }
+        }
 
-            if (IElement* e = SimplifyRefractContainer(defaults)) {
-                element->attributes["default"] = e;
-            }
+        if (IElement* e = SimplifyRefractContainer(samples)) {
+            element->attributes["sample"] = e;
```
Comment (zdne) [inline src/RefractAST.cc]: see my comment above on using constants – this should be really defined just once.

File: src/RefractAST.cc [hunk 3]
```diff
@@ -315,36 +327,82 @@ namespace drafter
...
         }
 
-        if(!description.empty()) {
-            element->meta["description"] = IElement::Create(description);
+        if (IElement* e = SimplifyRefractContainer(defaults)) {
+            element->attributes["default"] = e;
```
Comment (zdne) [inline src/RefractAST.cc]: -ditto-

File: src/RefractAST.cc [hunk 4]
```diff
@@ -315,36 +327,82 @@ namespace drafter
...
     }
 
     template <typename T>
-    refract::IElement* RefractElementFromProperty(const mson::PropertyMember& property)
+    refract::MemberElement* RefractElementFromProperty(const mson::PropertyMember& property)
     {
-        refract::IElement* element = RefractElementFromValue<T>(property);
-        if (!property.name.literal.empty()) {
-            element->meta["name"] = refract::IElement::Create(property.name.literal);
+        //if (property.name.literal.empty()) {
```
Comment (zdne) [inline src/RefractAST.cc]: commented -out code without any explanation. Please either add a comment explaining why this is commented out and until when or remove it completely

File: src/RefractAST.cc [hunk 5]
```diff
@@ -315,36 +327,82 @@ namespace drafter
...
                 // FIXME: is there example for multiple variables?
                 throw std::logic_error("Multiple variables in property definition");
             }
-            element->meta["name"] = refract::IElement::Create(property.name.variable.values[0].literal);
-            element->attributes["variable"] = refract::IElement::Create(true);
+            element->set(property.name.variable.values.begin()->literal, value);
+            element->value.first->attributes["variable"] = refract::IElement::Create(true);
+            SetElementType(property.name.variable.typeDefinition, element->value.first);
+            // FIXME: 
+            // https://github.com/refractproject/refract-spec/blob/master/namespaces/mson-namespace.md#variable-property-name
```
Comment (zdne) [inline src/RefractAST.cc]: A better explanation in the code of why is this "FIXME" would be good

File: src/RefractAST.cc [hunk 6]
```diff
@@ -315,36 +327,82 @@ namespace drafter
...
         }
         else if (!property.name.variable.values.empty()) {
             if (property.name.variable.values.size() > 1) {
                 // FIXME: is there example for multiple variables?
                 throw std::logic_error("Multiple variables in property definition");
             }
-            element->meta["name"] = refract::IElement::Create(property.name.variable.values[0].literal);
-            element->attributes["variable"] = refract::IElement::Create(true);
+            element->set(property.name.variable.values.begin()->literal, value);
+            element->value.first->attributes["variable"] = refract::IElement::Create(true);
```
Comment (zdne) [inline src/RefractAST.cc]: constant `"variable"`

File: src/RefractAST.cc [hunk 7]
```diff
@@ -558,7 +614,7 @@ namespace drafter
         SetElementType(ds.typeDefinition, e);
 
         e->meta["id"] = IElement::Create(ds.name.symbol.literal);
-        e->meta["title"] = IElement::Create(ds.name.symbol.literal);
+        //e->meta["title"] = IElement::Create(ds.name.symbol.literal);
```
Comment (zdne) [inline src/RefractAST.cc]: What is this for. Add explanation why is this commented out or remove it.

File: src/refract/Element.h [hunk 8]
```diff
@@ -270,22 +270,26 @@ namespace refract
 
     struct MemberElement : Element<MemberElement, MemberElementTrait>
     {
-        void set(const std::string& name, IElement* element)
+        void set(const std::string& key, IElement* element)
         {
+            set(IElement::Create(key) ,element);
```
Comment (zdne) [inline src/refract/Element.h]: white space :cop: ` ,`

File: src/refract/SerializeVisitor.cc [hunk 9]
```diff
@@ -108,7 +108,19 @@ namespace refract
 
     void SerializeVisitor::visit(const MemberElement& e)
     {
-        throw Deprecated("MemberElement is DEPRECATED use Meta[\"name\"] instead");
+        sos::Object object;
+        if(e.value.first) {
+            SerializeVisitor s;
+            s.visit(*e.value.first);
+            object.set("key",s.get());
```
Comment (zdne) [inline src/refract/SerializeVisitor.cc]: whitespace :cop:

**Comment by zdne** [comment]
@klokane looks great! I have some comment regarding defining and reusing contacts otherwise it is good to merge.

File: src/RefractAST.cc (hunk 5 above)
Comment (klokane) [inline src/RefractAST.cc]: it is explained few lines below in same comment - line no 373
In spec is used "typeAttributes: sample"
But in discussion we propose to remove "sample && defalut" from "typeAttributes"

While i move this explaining text uder FIXME, will be OK?

Comment (zdne) [inline src/RefractAST.cc]: @klokane OK but I do not understand why is this linking to the "Variable Property Name". Neither I do understand what should be fixed. Can you please elaborate ?

**Pull request merged**